clang/lib/Sema/SemaBounds.cpp

```diff
@@ -3998,17 +4027,28 @@ namespace {
       // starts at the beginning of a declaration T v = e, then extra
       // diagnostics may be emitted for T.
       SourceLocation Loc = St->getBeginLoc();
-      if (isa<DeclStmt>(St))
+      SourceRange SrcRange = St->getSourceRange();
+      auto BDCType = Sema::BoundsDeclarationCheck::BDC_Statement;
+      if (isa<DeclStmt>(St)) {
         Loc = V->getLocation();
+        BDCType = Sema::BoundsDeclarationCheck::BDC_Initialization;
+      }
+
+      // Find the assignment (if it exists) to blame for the error or warning.
+      auto It = State.BlameAssignments.find(V);
```

Comment:
Can this logic move into a helper method that sets a SourceLocation, SourceRange, and BDCType based on `St` and `BlameAssignments`?